### PR TITLE
Feature/90 known types

### LINKED ISSUE #90: IEnumerable<object> Generated for known types
### Motivation

When generating a class for a known content type that has embedded content types they get generated with incorrect types of "object" instead of the strongly typed class. Additionally, when a constraint is "Exactly 1", the object still gets generated as an IEnumerable (implying more than 1).

Here is an example:
image

This got generated as follows:

```cs
public IEnumerable<object> NavigationTemplate { get; set; }
```

It should be expected that this get generated as this:

```cs
public NavigationTemplate NavigationTemplate { get; set; }
```

### Proposed solution

Ideally the .NET code generator could differentiate between Exactly 1 and other values, and more importantly be bound to the correct type.

At a minimum it should have been able to figure out that it should have been an IEnumerable&lt;NavigationType&gt; instead of IEnumerable&lt;object&gt;

### PR DESCRIPTION
**Pull request by Sevitas**
### Motivation

Fixes #90. Scenarios 1 and 2.

### Checklist

- [x] Code follows coding conventions held in this repo
- [x] Automated tests have been added
- [x] Tests are passing
- [x] Docs have been updated (if applicable)
- [x] Temporary settings (e.g. variables used during development and testing) have been reverted to defaults

### How to test

Test that models are generated for senarios 1 and 2 and everything else works without a change.

### CODE REVIEW

**Comment by Simply007** [comment]
I will redo the review and wait for tracking header support with that.

**Comment by Simply007** [comment]
Mind the https://github.com/alanta/Kontent.Statiq/issues/21

**Review by Simply007 — CHANGES_REQUESTED** [review]
* I am getting the error: `Warning: Skipping unknown Content Element type 'Subpages'. (Content Type: 'Homepage', Element Codename: 'subpages').`
-> so subpages element is not supported, but it should be => #127 
* I am getting compilation error for this project "6d7f2b4b-9432-012e-f6be-c9feb74c3912" -> invited you inthere
* Add documentation -> it was hard to find right arguments for command line

I used this exe file -> direct build from the branch.
[KontentModelGenerator-win-x64-net6.0.zip](https://github.com/kontent-ai/model-generator-net/files/10508574/KontentModelGenerator-win-x64-net6.0.zip)

**Comment by Simply007** [comment]
Please adjust https://github.com/kontent-ai/model-generator-net/blob/feature/90_known_types/src/Kontent.Ai.ModelGenerator/appSettings.json to respect new options and cover related tests if necessary.

**Comment by Simply007** [comment]
I would emphasize the Api key with the example for the section: https://github.com/kontent-ai/model-generator-net/tree/feature/90_known_types#customizing-models---extended-delivery-models

**Comment by Simply007** [comment]
By running `.\KontentModelGenerator.exe -p ABC -e true -k XYZ` for project `6d7f2b4b-9432-012e-f6be-c9feb74c3912`.

I got also got for `*.Generated. inheritance from `IContentItem`.

**Review by Simply007 — CHANGES_REQUESTED** [review]
Casing problem

By running `.\KontentModelGenerator.exe -p ABC -e true -k XYZ` for project `6d7f2b4b-9432-012e-f6be-c9feb74c3912` (@Sevitas  - you are invited) => it is because faulty casing -> the original property is `JustMessages` -> the typed one is JustmessagesTypes and trying to load from Justmessages (see the `m` vs. M).

---

Opt in for `IContentItem`

1. I also got for `*.Generated`. inheritance from `IContentItem`. I would add this as non-default behavior. => I would extend the `structuredmodel` argument with another enum flag i.e. `linkedItems` (I am open to another name) because it probably (not sure) applies on the subPages element too.

1. Introduce `IContentItem` into the DAPI SDK Abstractions (I think the minor release is fine) and do not generate it.

File: README.md [hunk 1]
```diff
@@ -101,29 +106,45 @@ namespace KontentAiModels
...
-public partial class Home
+public partial class Home : IContentItem
 {
-    public IEnumerable<object> LinkedContentItems { get; set; }
+    public const string SingleAllowedTypeSingleLinkedContentItemCodename = "single_allowed_type_single_linked_content_item";
+    public const string SingleAllowedTypeMultiLinkedContentItemsCodename = "single_allowed_type_multi_linked_content_items";
+    public const string MultiAllowedTypesSingleLinkedContentItemCodename = "multi_allowed_types_single_linked_content_item";
+    public const string MultiAllowedTypesMultiLinkedContentItemsCodename = "multi_allowed_types_multi_linked_content_items";
+
+    // Allowed Content Types == "Article" && Limit number of items == 1
```
Comment (Simply007) [inline README.md:126]: I would extend this comment for not exactly and at most.

File: README.md [hunk 2]
```diff
@@ -47,16 +47,21 @@ To learn how to generate executables for your favorite target platform, follow t
...
+
+#### Structured model
+
+Generates `IRichTextContent` instead of `string` for rich-text elements or `IDateTimeContent` instead of `DateTime?` for date-time elements. This enables utilizing [structured rich-text rendering](https://github.com/kontent-ai/delivery-sdk-net/blob/master/docs/customization-and-extensibility/structured-models/structured-models-rendering.md) [structured date-time rendering](https://github.com/kontent-ai/delivery-sdk-net/blob/master/docs/customization-and-extensibility/structured-models/structured-models-rendering.md). For *modular content elements* and *subpages* generates [`IContentItem`](https://github.com/kontent-ai/delivery-sdk-net/blob/master/Kontent.Ai.Delivery.Abstractions/ContentItems/IContentItem.cs) instead of `object`, the model also implements the interface `IContentItem`.  Allowed values [`RichText`, `DateTime`, `True`, `ModularContent`], as a separator you should use `,`. ⚠️ `True` parameter is **obsolete** and interprets the same value as `RichText`.
```
Comment (Simply007) [inline README.md:64]: I would move info about modular content to the `structured-models-rendering.md` to keep it consistent.

**Review by Simply007 — CHANGES_REQUESTED** [review]
* rename the `extendeddelivermodels` - it is delivery everywhere we use it
* I am getting a warnings fron `self-contained.ps1`

  ```sh
  C:\projects\model-generator-net\src\Kontent.Ai.ModelGenerator\ServiceCollectionExtensions.cs(24,9): warning IL2026: Using member 'Microsoft.Extensions.Configur
  ation.ConfigurationBinder.Bind(IConfiguration, Object)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. C
  annot statically analyze the type of instance so its members may be trimmed. [C:\projects\model-generator-net\src\Kontent.Ai.ModelGenerator\Kontent.Ai.ModelGen 
  erator.csproj]
  C:\projects\model-generator-net\src\Kontent.Ai.ModelGenerator\Options\ArgHelpers.cs(97,32): warning IL2070: 'this' argument does not satisfy 'DynamicallyAccess 
  edMemberTypes.PublicProperties' in call to 'System.Type.GetProperties()'. The parameter 'type' of method 'Kontent.Ai.ModelGenerator.Options.ArgHelpers.ProgramO 
  ptionsData.ProgramOptionsData(Type, String)' does not have matching annotations. The source value must declare at least the same requirements as those declared 
   on the target location it is assigned to. [C:\projects\model-generator-net\src\Kontent.Ai.ModelGenerator\Kontent.Ai.ModelGenerator.csproj]
  C:\projects\model-generator-net\src\Kontent.Ai.ModelGenerator\Options\ArgHelpers.cs(105,32): warning IL2070: 'this' argument does not satisfy 'DynamicallyAcces 
  sedMemberTypes.PublicProperties' in call to 'System.Type.GetProperties()'. The parameter 'optionsType' of method 'Kontent.Ai.ModelGenerator.Options.ArgHelpers. 
  ProgramOptionsData.ProgramOptionsData(Type, Type, String)' does not have matching annotations. The source value must declare at least the same requirements as  
  those declared on the target location it is assigned to. [C:\projects\model-generator-net\src\Kontent.Ai.ModelGenerator\Kontent.Ai.ModelGenerator.csproj]       
  C:\projects\model-generator-net\src\Kontent.Ai.ModelGenerator\Program.cs(38,13): warning IL2026: Using member 'Microsoft.Extensions.DependencyInjection.Options 
  ConfigurationServiceCollectionExtensions.Configure<TOptions>(IServiceCollection, IConfiguration)' which has 'RequiresUnreferencedCodeAttribute' can break funct 
  ionality when trimming application code. TOptions's dependent types may have their members trimmed. Ensure all required members are preserved. [C:\projects\mod 
  el-generator-net\src\Kontent.Ai.ModelGenerator\Kontent.Ai.ModelGenerator.csproj]
    Kontent.Ai.ModelGenerator -> C:\projects\model-generator-net\src\Kontent.Ai.ModelGenerator\bin\Release\net6.0\linux-x64\KontentModelGenerator.dll
    Optimizing assemblies for size. This process might take a while.
  C:\projects\model-generator-net\src\Kontent.Ai.ModelGenerator\Program.cs(38,13): Trim analysis warning IL2026: Kontent.Ai.ModelGenerator.Program.<Main>d__0.Mov
  eNext(): Using member 'Microsoft.Extensions.DependencyInjection.OptionsConfigurationServiceCollectionExtensions.Configure<TOptions>(IServiceCollection, IConfig
  uration)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. TOptions's dependent types may have their membe 
  rs trimmed. Ensure all required members are preserved. [C:\projects\model-generator-net\src\Kontent.Ai.ModelGenerator\Kontent.Ai.ModelGenerator.csproj]
  C:\projects\model-generator-net\src\Kontent.Ai.ModelGenerator\Options\ArgHelpers.cs(105,13): Trim analysis warning IL2070: Kontent.Ai.ModelGenerator.Options.Ar 
  gHelpers.ProgramOptionsData.ProgramOptionsData(Type, Type, String): 'this' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicProperties' in call  
  to 'System.Type.GetProperties()'. The parameter 'optionsType' of method 'Kontent.Ai.ModelGenerator.Options.ArgHelpers.ProgramOptionsData.ProgramOptionsData(Typ 
  e, Type, String)' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it  
  is assigned to. [C:\projects\model-generator-net\src\Kontent.Ai.ModelGenerator\Kontent.Ai.ModelGenerator.csproj]
  C:\projects\model-generator-net\src\Kontent.Ai.ModelGenerator\Options\ArgHelpers.cs(97,13): Trim analysis warning IL2070: Kontent.Ai.ModelGenerator.Options.Arg 
  Helpers.ProgramOptionsData.ProgramOptionsData(Type, String): 'this' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicProperties' in call to 'Sys 
  tem.Type.GetProperties()'. The parameter 'type' of method 'Kontent.Ai.ModelGenerator.Options.ArgHelpers.ProgramOptionsData.ProgramOptionsData(Type, String)' do 
  es not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to. [C:\ 
  projects\model-generator-net\src\Kontent.Ai.ModelGenerator\Kontent.Ai.ModelGenerator.csproj]
  C:\projects\model-generator-net\src\Kontent.Ai.ModelGenerator\ServiceCollectionExtensions.cs(24,9): Trim analysis warning IL2026: Kontent.Ai.ModelGenerator.Ser 
  viceCollectionExtensions.LoadOptionsConfiguration(IServiceCollection, IConfiguration, String): Using member 'Microsoft.Extensions.Configuration.ConfigurationBi 
  nder.Bind(IConfiguration, Object)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. Cannot statically anal 
  yze the type of instance so its members may be trimmed. [C:\projects\model-generator-net\src\Kontent.Ai.ModelGenerator\Kontent.Ai.ModelGenerator.csproj]        
  ILLink : Trim analysis warning IL2026: System.Data.DataSet.System.Xml.Serialization.IXmlSerializable.GetSchema(): Using member 'System.Data.DataSet.WriteXmlSch 
  ema(DataSet, XmlWriter)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. DataSet.GetSchema uses TypeDescr 
  iptor and XmlSerialization underneath which are not trimming safe. Members from serialized types may be trimmed if not referenced directly. [C:\projects\model- 
  generator-net\src\Kontent.Ai.ModelGenerator\Kontent.Ai.ModelGenerator.csproj]
  ILLink : Trim analysis warning IL2026: System.Data.DataSet.System.Xml.Serialization.IXmlSerializable.ReadXml(XmlReader): Using member 'System.Data.DataSet.Read 
  XmlSerializableInternal(XmlReader)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. DataSet.ReadXml uses  
  XmlSerialization underneath which is not trimming safe. Members from serialized types may be trimmed if not referenced directly. [C:\projects\model-generator-n 
  et\src\Kontent.Ai.ModelGenerator\Kontent.Ai.ModelGenerator.csproj]
  ILLink : Trim analysis warning IL2026: System.Data.DataSet.System.Xml.Serialization.IXmlSerializable.WriteXml(XmlWriter): Using member 'System.Data.DataSet.Wri 
  teXmlInternal(XmlWriter)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. DataSet.WriteXml uses XmlSerial 
  ization underneath which is not trimming safe. Members from serialized types may be trimmed if not referenced directly. [C:\projects\model-generator-net\src\Ko 
  ntent.Ai.ModelGenerator\Kontent.Ai.ModelGenerator.csproj]
  ILLink : Trim analysis warning IL2026: System.Data.DataTable.System.Xml.Serialization.IXmlSerializable.GetSchema(): Using member 'System.Data.DataTable.GetXmlS 
  chema()' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. DataTable.GetSchema uses TypeDescriptor and XmlS 
  erialization underneath which are not trimming safe. Members from serialized types may be trimmed if not referenced directly. [C:\projects\model-generator-net\ 
  src\Kontent.Ai.ModelGenerator\Kontent.Ai.ModelGenerator.csproj]
  ILLink : Trim analysis warning IL2026: System.Data.DataTable.System.Xml.Serialization.IXmlSerializable.ReadXml(XmlReader): Using member 'System.Data.DataTable. 
  ReadXmlSerializableInternal(XmlReader)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. DataTable.ReadXml 
   uses XmlSerialization underneath which is not trimming safe. Members from serialized types may be trimmed if not referenced directly. [C:\projects\model-gener 
  ator-net\src\Kontent.Ai.ModelGenerator\Kontent.Ai.ModelGenerator.csproj]
  ILLink : Trim analysis warning IL2026: System.Data.DataTable.System.Xml.Serialization.IXmlSerializable.WriteXml(XmlWriter): Using member 'System.Data.DataTable 
  .WriteXmlInternal(XmlWriter)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. DataTable.WriteXml uses Xml 
  Serialization underneath which is not trimming safe. Members from serialized types may be trimmed if not referenced directly. [C:\projects\model-generator-net\ 
  src\Kontent.Ai.ModelGenerator\Kontent.Ai.ModelGenerator.csproj]
  C:\Users\OndrejChrastina\.nuget\packages\anglesharp\0.17.1\lib\netstandard2.0\AngleSharp.dll : warning IL2104: Assembly 'AngleSharp' produced trim warnings. Fo 
  r more information see https://aka.ms/dotnet-illink/libraries [C:\projects\model-generator-net\src\Kontent.Ai.ModelGenerator\Kontent.Ai.ModelGenerator.csproj]  
  C:\Users\OndrejChrastina\.nuget\packages\humanizer.core\2.14.1\lib\net6.0\Humanizer.dll : warning IL2104: Assembly 'Humanizer' produced trim warnings. For more
   information see https://aka.ms/dotnet-illink/libraries [C:\projects\model-generator-net\src\Kontent.Ai.ModelGenerator\Kontent.Ai.ModelGenerator.csproj]        
  C:\Users\OndrejChrastina\.nuget\packages\kontent.ai.delivery\17.4.0\lib\net6.0\Kontent.Ai.Delivery.dll : warning IL2104: Assembly 'Kontent.Ai.Delivery' produce 
  d trim warnings. For more information see https://aka.ms/dotnet-illink/libraries [C:\projects\model-generator-net\src\Kontent.Ai.ModelGenerator\Kontent.Ai.Mode 
  lGenerator.csproj]
  C:\Users\OndrejChrastina\.nuget\packages\kontent.ai.management\4.4.0\lib\net6.0\Kontent.Ai.Management.dll : warning IL2104: Assembly 'Kontent.Ai.Management' pr 
  oduced trim warnings. For more information see https://aka.ms/dotnet-illink/libraries [C:\projects\model-generator-net\src\Kontent.Ai.ModelGenerator\Kontent.Ai 
  .ModelGenerator.csproj]
  C:\Users\OndrejChrastina\.nuget\packages\microsoft.codeanalysis.common\4.5.0\lib\netcoreapp3.1\Microsoft.CodeAnalysis.dll : warning IL2104: Assembly 'Microsoft 
  .CodeAnalysis' produced trim warnings. For more information see https://aka.ms/dotnet-illink/libraries [C:\projects\model-generator-net\src\Kontent.Ai.ModelGen 
  erator\Kontent.Ai.ModelGenerator.csproj]
  C:\Users\OndrejChrastina\.nuget\packages\microsoft.codeanalysis.csharp.workspaces\4.5.0\lib\netcoreapp3.1\Microsoft.CodeAnalysis.CSharp.Workspaces.dll : warnin 
  g IL2104: Assembly 'Microsoft.CodeAnalysis.CSharp.Workspaces' produced trim warnings. For more information see https://aka.ms/dotnet-illink/libraries [C:\proje 
  cts\model-generator-net\src\Kontent.Ai.ModelGenerator\Kontent.Ai.ModelGenerator.csproj]
  C:\Users\OndrejChrastina\.nuget\packages\microsoft.codeanalysis.workspaces.common\4.5.0\lib\netcoreapp3.1\Microsoft.CodeAnalysis.Workspaces.dll : warning IL210 
  4: Assembly 'Microsoft.CodeAnalysis.Workspaces' produced trim warnings. For more information see https://aka.ms/dotnet-illink/libraries [C:\projects\model-gene 
  rator-net\src\Kontent.Ai.ModelGenerator\Kontent.Ai.ModelGenerator.csproj]
  C:\Users\OndrejChrastina\.nuget\packages\newtonsoft.json\13.0.1\lib\netstandard2.0\Newtonsoft.Json.dll : warning IL2104: Assembly 'Newtonsoft.Json' produced tr 
  im warnings. For more information see https://aka.ms/dotnet-illink/libraries [C:\projects\model-generator-net\src\Kontent.Ai.ModelGenerator\Kontent.Ai.ModelGen 
  erator.csproj]
  C:\Users\OndrejChrastina\.nuget\packages\scrutor\4.2.0\lib\net6.0\Scrutor.dll : warning IL2104: Assembly 'Scrutor' produced trim warnings. For more information 
   see https://aka.ms/dotnet-illink/libraries [C:\projects\model-generator-net\src\Kontent.Ai.ModelGenerator\Kontent.Ai.ModelGenerator.csproj]
  C:\Users\OndrejChrastina\.nuget\packages\system.composition.convention\6.0.0\lib\net6.0\System.Composition.Convention.dll : warning IL2104: Assembly 'System.Co 
  mposition.Convention' produced trim warnings. For more information see https://aka.ms/dotnet-illink/libraries [C:\projects\model-generator-net\src\Kontent.Ai.M 
  odelGenerator\Kontent.Ai.ModelGenerator.csproj]
  C:\Users\OndrejChrastina\.nuget\packages\system.composition.hosting\6.0.0\lib\net6.0\System.Composition.Hosting.dll : warning IL2104: Assembly 'System.Composit 
  ion.Hosting' produced trim warnings. For more information see https://aka.ms/dotnet-illink/libraries [C:\projects\model-generator-net\src\Kontent.Ai.ModelGener 
  ator\Kontent.Ai.ModelGenerator.csproj]
  C:\Users\OndrejChrastina\.nuget\packages\system.composition.typedparts\6.0.0\lib\net6.0\System.Composition.TypedParts.dll : warning IL2104: Assembly 'System.Co 
  mposition.TypedParts' produced trim warnings. For more information see https://aka.ms/dotnet-illink/libraries [C:\projects\model-generator-net\src\Kontent.Ai.M 
  odelGenerator\Kontent.Ai.ModelGenerator.csproj]
  ```

* I would use `FirstOrDefault` also for `AtMost` = 1 linked items - what do you think?

File: README.md (hunk 1 above)
Comment (Sevitas) [reply README.md:126]: I think it might be better to introduce separate property

File: README.md
Comment (Simply007) [inline README.md:1]: ### CLI Syntax

I was not able to use this CLI syntax for arguments:
* -e=true
* -e True
* -e=True

**Review by Simply007 — CHANGES_REQUESTED** [review]

File: README.md [hunk 3]
```diff
@@ -47,16 +47,20 @@ To learn how to generate executables for your favorite target platform, follow t
...
+| `-t`      |           `--withtypeprovider`           |  False   |      `true`       |                                          Indicates whether the `CustomTypeProvider` class should be generated (see [Customizing the strong-type binding logic](https://github.com/kontent-ai/delivery-sdk-net/blob/master/docs/customization-and-extensibility/strongly-typed-models.md#customizing-the-strong-type-binding-logic) for more info)                                              |
+| `-s`      |           `--structuredmodel`            |  False   |      `null`       | Allowed values [`RichText`, `DateTime`, `True`, `ModularContent`], as a separator you should use `,`. ⚠️ `True` parameter is **obsolete** and interprets the same value as `RichText`. For further details see [structured models rendering](https://github.com/kontent-ai/delivery-sdk-net/blob/master/docs/customization-and-extensibility/structured-models/structured-models-rendering.md) |
```
[the remark this answers is not in the document]
Comment (Simply007) [reply README.md:57]: Let's keep it this way.

**Review by Simply007 — APPROVED** [review]
LGTM
 tested locally and from nuget

**Pull request merged**